onnx_chainer/export.py

```diff
@@ -0,0 +1,509 @@
+                    temp_var = chainer.as_variable(inputs[i])
+                    func_inodes[i] = temp_var.node
+                    self.retain_inputs.append(temp_var)
+                else:
+                    if id(referenced_var) not in self.link_inputs:
+                        # This variable is created within link forward, outside
+                        # of function node. To avoid to lose reference out
+                        # of the forward, retain the variable.
+                        self.retain_inputs.append(referenced_var)
+            self.replaced_inputs.append((_self, _self.inputs))
+            _self.inputs = tuple(func_inodes)
+            return ret
+        self.hooked_apply = hooked_apply
+
+    def _extract_inputs(self, args):
+        # Retain only chainer.Variable (and its collection)
+        # Other type args are ignored and not checked instance IDs
+        # If these variable are used in FunctionNode, they will be retained
+        ret = set()
+        if isinstance(args, chainer.Variable):
+            ret.add(id(args))
+        elif isinstance(args, (list, tuple)):
+            for arg in args:
+                ret |= self._extract_inputs(arg)
+        elif isinstance(args, dict):
+            for arg in args.values():
+                ret |= self._extract_inputs(arg)
+        return ret
+
+    def forward_preprocess(self, args):
+        self.link_inputs |= self._extract_inputs(args.args)
+        self.link_inputs |= self._extract_inputs(args.kwargs)
+
+    def forward_postprocess(self, args):
+        self.link_inputs.clear()
+
+    def __enter__(self):
+        chainer.function_node.FunctionNode.apply = self.hooked_apply
+        return super().__enter__()
+
+    def __exit__(self, *exc_details):
+        chainer.function_node.FunctionNode.apply = self.org_apply
+        for _self, inputs in self.replaced_inputs:
+            _self.inputs = inputs
+        super().__exit__(*exc_details)
+
+
+def export(model, args, filename=None, export_params=True,
+           graph_name='Graph', save_text=False, opset_version=None,
+           input_names=None, output_names=None, train=False,
+           return_named_inout=False, external_converters=None,
+           external_opset_imports=None, input_shapes=None):
+    """Export function for chainer.Chain in ONNX format.
+
+    This function performs a forward computation of the given
+    :class:`~chainer.Chain`, ``model``, by passing the given arguments ``args``
+    directly. It means, the output :class:`~chainer.Variable` object ``y`` to
+    make the computational graph will be created by:
+
+    ``y = model(*args)``
+
+    ``external_converters`` and ``external_opset_imports`` are for external
+    custom operator. When some ~chainer.FunctionNode are expected to convert to
+    own customized operator, set converter function with ~chainer.FunctionNode
+    name.
+
+    >>> import onnx
+    >>> def custom_converter(param):
```

Comment:
doctest failed here.  It seems `...` (in place of `>>>`) should be used on continuation lines.  E.g.
```
    >>> def custom_converter(param):
    ...     return onnx.helper.make_node(
    ...         'CustomizedRelu', param.input_names, param.output_names,
    ...         domain='chainer'),
```

Reply:
Thanks, got it! I fixed them.